### PR TITLE
#9 Upgrade Electron to v24

### PR DESCRIPTION
**Pull request by viktor44**
### What is this PR

Upgraded to Electron 24

### Fixed issues

* #165
* #161
* #151
* #143
* #108
* #9

### CODE REVIEW

File: app/plugins/webview-preload.js [hunk 1]
```diff
@@ -28,7 +27,8 @@ if (micromatch.isMatch(window.location.origin, originsAllowed) || protocolsAllow
       });
     });
     setTimeout(() => {
-      ipcRenderer.sendTo(workerWebContentsId, 'bx-api-perform', channel, payload);
+      ipcRenderer.invoke('get-worker-contents-id')
```
Comment (magne4000) [inline app/plugins/webview-preload.js:30]: Perhaps store the result of this call so we just have to call it once per process.

File: app/plugins/webview-preload.js [hunk 2]
```diff
@@ -40,7 +40,8 @@ if (micromatch.isMatch(window.location.origin, originsAllowed) || protocolsAllow
       });
     });
     setTimeout(() => {
-      ipcRenderer.sendTo(workerWebContentsId, 'bx-api-subscribe', channel);
+      ipcRenderer.invoke('get-worker-contents-id')
```
Comment (magne4000) [inline app/plugins/webview-preload.js:43]: IDEM

File: app/services/main/boot.ts [hunk 3]
```diff
@@ -3,9 +3,18 @@
 /*
 ** This file purpose is to handle startup side effects in main process
 */
+import { app, ipcMain } from 'electron';
 
 import { startSessionsListening } from '../api/sessions';
 
 export default () => {
+  
+  require('stream-electron-ipc');
```
Comment (magne4000) [inline app/services/main/boot.ts:12]:
```suggestion
  // make sure that `stream-electron-ipc` is call early
  require('stream-electron-ipc');
```

**Comment by magne4000** [comment]
Migrated to node@18, the build version works, but `yarn run dev` does not, because I cannot forward `--openssl-legacy-provider` to the dev process (Electron forbids it). See https://github.com/webpack/webpack/issues/14532

To fix that, we have multiple solutions:
- Webpack would need monkeypatching to replace all `md4` usage to `xxhash64`
- Upgrade to `webpack@5`
- Upgrade to `electron-forge`

All those solutions are not trivial. Finding a way to force/monkeypatch `output.hashFunction: "xxhash64"` on webpack AND all its plugins would probably be the simplest solution.

I will not have time to continue working on this for the next few weeks, so feel free to take over this subject.

**Comment by viktor44** [comment]
But I have the same test error with Node 18

```
 FAIL  test/jest/bang/search/test-search.ts
  ● Test suite failed to run

    1:undefined
    undefined

      at Object.reject (node_modules/@jest-runner/rpc/build/RPCProcess.js:139:18)
      at Server.emit (node_modules/event-pubsub/es5.js:74:21)
      at Server.gotData (node_modules/node-ipc/dao/socketServer.js:194:14)
```

hence Node upgrade won't help with this problem

**Comment by magne4000** [comment]
The test is fixed, it was due to some weird import issue :man_shrugging:

Now only remains to fix `yarn run dev`.

**Comment by viktor44** [comment]
I also moved your yesterday commits to separate branch
https://github.com/viktor44/station-desktop-app/tree/feature/electron_24_node_18

**Comment by magne4000** [comment]
Electron 24 is tied to Node 18, so migrating one without the other is not a complete PR and will lead to unexpected behaviour.

**Comment by magne4000** [comment]
@viktor44 Good to merge on my side. Ok for you too?

File: app/plugins/webview-preload.js (hunk 1 above)
Comment (viktor44) [reply app/plugins/webview-preload.js:30]: > Perhaps store the result of this call so we just have to call it once per process.

I have doubts about this suggestion
Should we create a module for this purpose?
and then use it like 

```javascript
  const id = require('some-new-module').workerWebContentsId
```

What is the right place to put it?
Is this `worker-id` available in any moment of time and code? What if we ask for it too early?

Comment (magne4000) [reply app/plugins/webview-preload.js:30]: A module seems a little overkill, I was more thinking just storing the result of the call in a variable in the file itself

Comment (viktor44) [reply app/plugins/webview-preload.js:30]: That was my first version :)
Like this

```javascript
  const workerWebContentsIdPromise = ipcRenderer.invoke('get-worker-contents-id');

  const sendPerformToProxy = (channel, payload) => {
    const p = new Promise(resolve => {
      ipcRenderer.once(`bx-api-perform-response-${channel}`, (_, result) => {
        resolve(result);
      });
    });

    setTimeout(() => {
        workerWebContentsIdPromise
            .then(workerWebContentsId => ipcRenderer.sendTo(workerWebContentsId, 'bx-api-perform', channel, payload));
    }, 1);
    return p;
  };
```

But the same concern about requesting it too early.
And `setTimeout` here looks like brute workaround, so I decided not to touch it and to keep as is.

**Pull request merged**